### PR TITLE
Use database filters to ensure execution event phase transitions are valid

### PR DESCRIPTION
**Pull request by katrogan**
# TL;DR
Use database filters to ensure execution event phase transitions are valid.

## Type
 - [x] Bug Fix
 - [ ] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [x] Unit tests added
 - [x] Code documentation added
 - [x] Any pending items have an associated Issue

## Complete description
 _How did you fix the bug, make the feature etc. Link to any design docs etc_

## Tracking Issue

fixes https://github.com/flyteorg/flyte/issues/2186

## Follow-up issue
_NA_

### CODE REVIEW

File: pkg/manager/impl/executions/phase_filters.go [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+		return nil, errors.NewFlyteAdminErrorf(codes.InvalidArgument, "Unrecognized execution entity [%+v] for non-terminal filters", entity)
+	}
+}
+
+func newSchedulingFilter() (common.InlineFilter, error) {
+	return common.NewRepeatedValueFilter(common.Execution, common.ValueIn, "phase", SchedulingWorkflowPhases)
+}
+
+func GetUpdateExecutionFilters(eventPhase core.WorkflowExecution_Phase) (filters []common.InlineFilter, err error) {
+	// With the exeception of queued events, it's never acceptable to move into the same phase.
```
Comment (EngHabu) [inline pkg/manager/impl/executions/phase_filters.go:57]: Hmmm I'm not sure this is accurate. If propeller picks up a WF, sends running event then fails to write the CRD back to ApiServer. It'll attempt to do the same operation again... why is this a bad thing not just an idempotent operation?

**Review by EngHabu — COMMENTED** [review]
I've one question..... but as an overall design question, it feels like we are trying to replicate what RevisionVersion does for CRDs for WFs and Tasks using Phases... 
In the case of K8s, it allows any kind of update as long as the caller knows what they are doing (i.e.  they have the latest version of the CRD)...

File: pkg/manager/impl/executions/phase_filters.go (hunk 1 above)
Comment (katrogan) [reply pkg/manager/impl/executions/phase_filters.go:57]: oh interesting, the current implementation doesn't let us update to an existing phase. i thought propeller swallowed the already exists in that case e.g. https://github.com/flyteorg/flyteadmin/blob/master/pkg/manager/impl/execution_manager.go#L1183
andhttps://github.com/flyteorg/flytepropeller/blob/c016dabbfef6037bead59590b42326dabe89f957/pkg/controller/workflow/executor.go#L261

perhaps propeller should also be modified to swallow the NotFound when RowsAffected == 0 gets triggered as part of meeting the new phase conditions. but perhaps that is getting too finnicky

**Comment by katrogan** [comment]
Re: your point above, I agree that this verification is a little clunky but I worry doing a get and then update will incur a performance penalty. what do you think? if we pursue that i don't think the issue this change addresses is fundamentally significant enough to introduce that design in the short-term :\

**Pull request closed without merging**